File: backend/src/services/routeService.js
const fetch = require('node-fetch');

const OSRM_API = process.env.OSRM_API_URL || 'https://router.project-osrm.org';
const NOMINATIM_API = process.env.NOMINATIM_API_URL || 'https://nominatim.openstreetmap.org';

class RouteService {
  /**
   * Get a route between two points using OSRM
   * @param {number} fromLon - Start longitude
   * @param {number} fromLat - Start latitude
   * @param {number} toLon - End longitude
   * @param {number} toLat - End latitude
   * @param {Array} avoidPoints - Array of [lon, lat] points to avoid
   * @returns {Object} Route data with geometry, duration, distance
   */
  async getRoute(fromLon, fromLat, toLon, toLat, avoidPoints = []) {
    try {
      // Build waypoints string
      let coordinates = `${fromLon},${fromLat};${toLon},${toLat}`;
      
      const url = `${OSRM_API}/route/v1/driving/${coordinates}?overview=full&geometries=geojson&steps=true&alternatives=true`;
      
      const response = await fetch(url, {
        headers: {
          'User-Agent': 'CrowdNav/1.0',
        },
      });
      
      if (!response.ok) {
        throw new Error(`OSRM API error: ${response.status}`);
      }
      
      const data = await response.json();
      
      if (data.code !== 'Ok' || !data.routes || data.routes.length === 0) {
        throw new Error('No route found');
      }
      
      // If we have avoid points, try to filter or re-route
      let bestRoute = data.routes[0];
      
      if (avoidPoints.length > 0 && data.routes.length > 1) {
        // Score each route by how far it stays from avoid points
        bestRoute = this._selectBestRoute(data.routes, avoidPoints);
      }
      
      return {
        geometry: bestRoute.geometry,
        duration: bestRoute.duration,
        distance: bestRoute.distance,
        steps: bestRoute.legs[0]?.steps || [],
        alternatives: data.routes.slice(1).map(r => ({
          geometry: r.geometry,
          duration: r.duration,
          distance: r.distance,
        })),
      };
    } catch (error) {
      console.error('Route generation error:', error);
      throw error;
    }
  }
  
  /**
   * Get route with intermediate waypoints to avoid problem areas
   */
  async getRouteAvoidingAreas(fromLon, fromLat, toLon, toLat, avoidAreas) {
    try {
      // Generate waypoints that route around avoid areas
      const waypoints = this._generateAvoidanceWaypoints(
        fromLon, fromLat, toLon, toLat, avoidAreas
      );
      
      let coordStr = `${fromLon},${fromLat}`;
      for (const wp of waypoints) {
        coordStr += `;${wp[0]},${wp[1]}`;
      }
      coordStr += `;${toLon},${toLat}`;
      
      const url = `${OSRM_API}/route/v1/driving/${coordStr}?overview=full&geometries=geojson&steps=true`;
      
      const response = await fetch(url, {
        headers: { 'User-Agent': 'CrowdNav/1.0' },
      });
      
      const data = await response.json();
      
      if (data.code !== 'Ok' || !data.routes?.length) {
        // Fallback to direct route
        return this.getRoute(fromLon, fromLat, toLon, toLat);
      }
      
      return {
        geometry: data.routes[0].geometry,
        duration: data.routes[0].duration,
        distance: data.routes[0].distance,
        steps: data.routes[0].legs?.flatMap(l => l.steps) || [],
        avoidedAreas: avoidAreas.length,
      };
    } catch (error) {
      console.error('Avoidance routing error:', error);
      return this.getRoute(fromLon, fromLat, toLon, toLat);
    }
  }
  
  /**
   * Geocode a place name to coordinates
   */
  async geocode(query) {
    try {
      const url = `${NOMINATIM_API}/search?q=${encodeURIComponent(query)}&format=json&limit=5&addressdetails=1`;
      
      const response = await fetch(url, {
        headers: { 'User-Agent': 'CrowdNav/1.0' },
      });
      
      const data = await response.json();
      
      return data.map(item => ({
        displayName: item.display_name,
        latitude: parseFloat(item.lat),
        longitude: parseFloat(item.lon),
        type: item.type,
        address: item.address,
      }));
    } catch (error) {
      console.error('Geocoding error:', error);
      throw error;
    }
  }
  
  /**
   * Reverse geocode coordinates to an address
   */
  async reverseGeocode(latitude, longitude) {
    try {
      const url = `${NOMINATIM_API}/reverse?lat=${latitude}&lon=${longitude}&format=json`;
      
      const response = await fetch(url, {
        headers: { 'User-Agent': 'CrowdNav/1.0' },
      });
      
      const data = await response.json();
      
      return {
        displayName: data.display_name,
        address: data.address,
      };
    } catch (error) {
      console.error('Reverse geocoding error:', error);
      throw error;
    }
  }
  
  _selectBestRoute(routes, avoidPoints) {
    let bestRoute = routes[0];
    let bestScore = -Infinity;
    
    for (const route of routes) {
      let minDist = Infinity;
      const coords = route.geometry.coordinates;
      
      for (const avoid of avoidPoints) {
        for (const coord of coords) {
          const dist = this._haversineDistance(coord[1], coord[0], avoid[1], avoid[0]);
          minDist = Math.min(minDist, dist);
        }
      }
      
      // Score = distance from problems - route duration penalty
      const score = minDist - (route.duration / 60);
      
      if (score > bestScore) {
        bestScore = score;
        bestRoute = route;
      }
    }
    
    return bestRoute;
  }
  
  _generateAvoidanceWaypoints(fromLon, fromLat, toLon, toLat, avoidAreas) {
    const waypoints = [];
    const offsetDeg = 0.003; // ~300m offset
    
    for (const area of avoidAreas) {
      const aLon = area.coordinates[0];
      const aLat = area.coordinates[1];
      
      // Check if this area is roughly between start and end
      const minLon = Math.min(fromLon, toLon) - 0.01;
      const maxLon = Math.max(fromLon, toLon) + 0.01;
      const minLat = Math.min(fromLat, toLat) - 0.01;
      const maxLat = Math.max(fromLat, toLat) + 0.01;
      
      if (aLon >= minLon && aLon <= maxLon && aLat >= minLat && aLat <= maxLat) {
        // Add a waypoint that bypasses the problem area
        const bearing = Math.atan2(toLat - fromLat, toLon - fromLon);
        const perpBearing = bearing + Math.PI / 2;
        
        waypoints.push([
          aLon + offsetDeg * Math.cos(perpBearing),
          aLat + offsetDeg * Math.sin(perpBearing),
        ]);
      }
    }
    
    return waypoints;
  }
  
  _haversineDistance(lat1, lon1, lat2, lon2) {
    const R = 6371000;
    const dLat = (lat2 - lat1) * Math.PI / 180;
    const dLon = (lon2 - lon1) * Math.PI / 180;
    const a = Math.sin(dLat / 2) ** 2 +
      Math.cos(lat1 * Math.PI / 180) * Math.cos(lat2 * Math.PI / 180) *
      Math.sin(dLon / 2) ** 2;
    return R * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  }
}

module.exports = new RouteService();

File: backend/services/route_service.py
import math
import logging
import httpx
from config import OSRM_API_URL, NOMINATIM_API_URL

logger = logging.getLogger("crowdnav.route")

class RouteService:
    async def get_route(self, from_lon: float, from_lat: float, to_lon: float, to_lat: float, avoid_points: list = None):
        if avoid_points is None:
            avoid_points = []
            
        coordinates = f"{from_lon},{from_lat};{to_lon},{to_lat}"
        url = f"{OSRM_API_URL}/route/v1/driving/{coordinates}?overview=full&geometries=geojson&steps=true&alternatives=true"
        
        async with httpx.AsyncClient(timeout=10.0) as client:
            response = await client.get(url, headers={"User-Agent": "CrowdNav-Python/1.0"})
            if response.status_code != 200:
                raise RuntimeError(f"OSRM API error: {response.status_code}")
            
            data = response.json()
            if data.get("code") != "Ok" or not data.get("routes"):
                raise RuntimeError("No route found")
            
            routes = data["routes"]
            best_route = routes[0]
            
            if len(avoid_points) > 0 and len(routes) > 1:
                best_route = self._select_best_route(routes, avoid_points)
                
            steps = []
            if best_route.get("legs") and len(best_route["legs"]) > 0:
                steps = best_route["legs"][0].get("steps", [])
                
            alternatives = []
            for r in routes[1:]:
                alternatives.append({
                    "geometry": r.get("geometry"),
                    "duration": r.get("duration"),
                    "distance": r.get("distance"),
                })
                
            return {
                "geometry": best_route.get("geometry"),
                "duration": best_route.get("duration"),
                "distance": best_route.get("distance"),
                "steps": steps,
                "alternatives": alternatives,
            }

    async def get_route_avoiding_areas(self, from_lon: float, from_lat: float, to_lon: float, to_lat: float, avoid_areas: list):
        try:
            waypoints = self._generate_avoidance_waypoints(from_lon, from_lat, to_lon, to_lat, avoid_areas)
            
            coord_str = f"{from_lon},{from_lat}"
            for wp in waypoints:
                coord_str += f";{wp[0]},{wp[1]}"
            coord_str += f";{to_lon},{to_lat}"
            
            url = f"{OSRM_API_URL}/route/v1/driving/{coord_str}?overview=full&geometries=geojson&steps=true"
            
            async with httpx.AsyncClient(timeout=10.0) as client:
                response = await client.get(url, headers={"User-Agent": "CrowdNav-Python/1.0"})
                data = response.json()
                
                if data.get("code") != "Ok" or not data.get("routes"):
                    return await self.get_route(from_lon, from_lat, to_lon, to_lat)
                
                route = data["routes"][0]
                steps = []
                if route.get("legs"):
                    for leg in route["legs"]:
                        steps.extend(leg.get("steps", []))
                        
                return {
                    "geometry": route.get("geometry"),
                    "duration": route.get("duration"),
                    "distance": route.get("distance"),
                    "steps": steps,
                    "avoidedAreas": len(avoid_areas),
                }
        except Exception as e:
            logger.error(f"Avoidance routing error: {e}")
            return await self.get_route(from_lon, from_lat, to_lon, to_lat)

    async def geocode(self, query: str):
        url = f"{NOMINATIM_API_URL}/search?q={query}&format=json&limit=5&addressdetails=1"
        async with httpx.AsyncClient(timeout=10.0) as client:
            response = await client.get(url, headers={"User-Agent": "CrowdNav-Python/1.0"})
            data = response.json()
            results = []
            for item in data:
                results.append({
                    "displayName": item.get("display_name"),
                    "latitude": float(item.get("lat", 0)),
                    "longitude": float(item.get("lon", 0)),
                    "type": item.get("type"),
                    "address": item.get("address"),
                })
            return results

    async def reverse_geocode(self, latitude: float, longitude: float):
        url = f"{NOMINATIM_API_URL}/reverse?lat={latitude}&lon={longitude}&format=json"
        async with httpx.AsyncClient(timeout=10.0) as client:
            response = await client.get(url, headers={"User-Agent": "CrowdNav-Python/1.0"})
            data = response.json()
            return {
                "displayName": data.get("display_name"),
                "address": data.get("address"),
            }

    def _select_best_route(self, routes: list, avoid_points: list):
        best_route = routes[0]
        best_score = float("-inf")
        
        for route in routes:
            min_dist = float("inf")
            coords = route.get("geometry", {}).get("coordinates", [])
            
            for avoid in avoid_points:
                for coord in coords:
                    dist = self._haversine_distance(coord[1], coord[0], avoid[1], avoid[0])
                    min_dist = min(min_dist, dist)
                    
            score = min_dist - (route.get("duration", 0) / 60.0)
            if score > best_score:
                best_score = score
                best_route = route
                
        return best_route

    def _generate_avoidance_waypoints(self, from_lon: float, from_lat: float, to_lon: float, to_lat: float, avoid_areas: list):
        waypoints = []
        offset_deg = 0.003  # ~300m offset
        
        for area in avoid_areas:
            area_coords = area.get("coordinates", [0, 0])
            a_lon, a_lat = area_coords[0], area_coords[1]
            
            min_lon = min(from_lon, to_lon) - 0.01
            max_lon = max(from_lon, to_lon) + 0.01
            min_lat = min(from_lat, to_lat) - 0.01
            max_lat = max(from_lat, to_lat) + 0.01
            
            if min_lon <= a_lon <= max_lon and min_lat <= a_lat <= max_lat:
                bearing = math.atan2(to_lat - from_lat, to_lon - from_lon)
                perp_bearing = bearing + (math.pi / 2.0)
                waypoints.append([
                    a_lon + offset_deg * math.cos(perp_bearing),
                    a_lat + offset_deg * math.sin(perp_bearing),
                ])
                
        return waypoints

    def _haversine_distance(self, lat1: float, lon1: float, lat2: float, lon2: float) -> float:
        r = 6371000  # meters
        d_lat = math.radians(lat2 - lat1)
        d_lon = math.radians(lon2 - lon1)
        a = (math.sin(d_lat / 2.0) ** 2 +
             math.cos(math.radians(lat1)) * math.cos(math.radians(lat2)) *
             math.sin(d_lon / 2.0) ** 2)
        return r * 2.0 * math.atan2(math.sqrt(a), math.sqrt(1 - a))

route_service = RouteService()
